### PR TITLE
Finish Operator once it is unblocked

### PR DESCRIPTION
**Pull request by findepi**
Avoid calling `Operator#finish()` when the operator is not yet ready to
finish, as indicated by `Operator#isBlocked()`.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 1]
```diff
@@ -317,7 +317,7 @@ private void processNewSources()
                 }
 
                 // if current operator is finished...
-                if (current.isFinished()) {
+                if (current.isFinished() && !next.isFinished() && isBlocked(next).isDone()) {
```
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: I would document this contract about not calling `finish()` when operator `isBlocked()` in `Operator`'s java docs

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: `isBlocked` is treated as only a suggestion from the operator and I don't think we should we should be changing that.

Both of these changes are something that the operator need to code for so I don't think we should be changing this.

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: @dain the idea is that would simplify `Operator` implementation at no cost. Note that Driver won't call `Operator.addInput` when operator is blocked, so why not guarantee `Operator.finish()` won't be called when blocked either?

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: I can drop `&& !next.isFinished()` from PR, if you don't like this part of the change.

Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: @dain note also that `isBlocked(next).isDone()` is not the same as `operator.isBlocked().isDone()`. Difference lies in memory allocation (operator that is not _internally_ blocked can still be blocked from Driver's perspective if waits for memory). Calling `finish` in such state sounds like not-a-problem unless Operator allocates even more memory in finish (as `HashBuilderOperator` does).

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: Currently, the design of the interface between Drivers and Operators is that Operator is required to deal with it's state machine by itself.  The blocked future is simply treated as a suggestion from the Operator and the Driver is (and does) free to choose to ignore this suggestion.  If there is a operator that can is having problems with this, then it is a bug in the operator. 

We can talk about changing this design, but that is a pretty big undertaking (and in my mind without much benefit).  As for this specific change, I don't see why we would want to make this.  To me this is just masking over a problem in some other operator.

**Comment by findepi** [comment]
Thank you @dain for the explanation. By "changing this design [...] is a pretty big undertaking" you mean it would involve adjusting `finish()` methods in all operators to remove now-unnecessary checks?

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 1 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: @dain, but many (some?) operators relays on `isBlocked()` preventing from calling `addInput()` and maybe other methods. For example `PartitionedOutputOperator` and `TaskOutputOperator` both have following code:
```
    @Override
    public void addInput(Page page)
    {
        requireNonNull(page, "page is null");
        checkState(isBlocked().isDone(), "output is already blocked");
        (...)
    }
```

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: The only guaranteed method ordering is that before `addInput` is called, `needsInput` will have returned true.  Any operator that depends on `isBlocked` has a bug.  `PartitionedOutputOperator` correctly implements this behavior as follows:
```java
    @Override
    public boolean needsInput()
    {
        return !finished && isBlocked().isDone();
    }

    @Override
    public void addInput(Page page)
    {
        requireNonNull(page, "page is null");
        checkState(isBlocked().isDone(), "output is already blocked");

        if (page.getPositionCount() == 0) {
            return;
        }

        page = pagePreprocessor.apply(page);
        blocked = partitionFunction.partitionPage(page);

        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
    }
```

**Comment by dain** [comment]
I really mean that if we are going to drastically change the design, then we should do a full design review which is a bigger undertaking.  For example, it would be nice to have an explicit state machine for operators so it would be easier to understand which parts of the finished are finished.

**Comment by pnowojski** [comment]
Ok, I get it now. This contract that you described could be documented somewhere in the comments.

**Comment by findepi** [comment]
@dain, if the intention is to deliver `finish()` even if operator is blocked, let's make it both: clear and reliable. #8186

**Pull request closed without merging**